### DISCUSSION TITLE
How cheap is construction?

### DISCUSSION THREAD

**Discussion by ThatXliner** [Rahix/avr-hal#598] [Q&A] [answered]
Should I only have 1 call of `arduino_hal::pins!` throughout the whole program? What if I want to do stuff on some pins in some functions? How do I pass it around? The Pins struct isn't `Copy` or `Clone`, and accessing a pin requires a borrow...

**Comment by ThatXliner** [comment, marked as answer]
Nevermind, just ~~borrow~~ move the Pin and return it when you're done

**Reply by Rahix** [reply]
To give some context:

The `Pins` struct is meant to be a singleton &mdash; only one is allowed to exist in your program.  It is designed this way because each pin only exists once in hardware.  By having only one `Pins` struct, each `Pin` also only exists once.  So by design, you can't ever perform conflicting operations on the same pin.

You already figured out the right way to work with this:  Either borrow or move the pins around as needed.  A borrowed pin cannot change mode while a moved pin can be reconfigured by the receiver.
